types/types.go

```diff
@@ -53,24 +56,50 @@ func NewWarpBlockInfo(header *types.Header, ethClient ethclient.Client) (*WarpBl
 			return nil, err
 		}
 	}
+	messages := make([]*WarpMessageInfo, len(logs))
+	for i, log := range logs {
+		warpLog, err := NewWarpMessageInfo(log)
+		if err != nil {
+			return nil, err
+		}
+		messages[i] = warpLog
+	}
+
 	return &WarpBlockInfo{
 		BlockNumber: header.Number.Uint64(),
-		WarpLogs:    logs,
+		Messages:    messages,
 	}, nil
 }
 
 // Extract the Warp message information from the raw log
-func NewWarpLogInfo(log types.Log) (*WarpLogInfo, error) {
+func NewWarpMessageInfo(log types.Log) (*WarpMessageInfo, error) {
 	if len(log.Topics) != 3 {
 		return nil, ErrInvalidLog
 	}
 	if log.Topics[0] != WarpPrecompileLogFilter {
 		return nil, ErrInvalidLog
 	}
+	unsignedMsgBytes, err := UnpackWarpMessage(log.Data)
+	if err != nil {
+		return nil, err
+	}
 
-	return &WarpLogInfo{
-		// BytesToAddress takes the last 20 bytes of the byte array if it is longer than 20 bytes
-		SourceAddress:    common.BytesToAddress(log.Topics[1][:]),
-		UnsignedMsgBytes: log.Data,
+	return &WarpMessageInfo{
+		SourceAddress:   common.BytesToAddress(log.Topics[1][:]),
+		UnsignedMessage: unsignedMsgBytes,
 	}, nil
 }
+
+func UnpackWarpMessage(unsignedMsgBytes []byte) (*avalancheWarp.UnsignedMessage, error) {
+	unsignedMsg, err := warp.UnpackSendWarpEventDataToMessage(unsignedMsgBytes)
+	if err != nil {
```

Comment:
Could you clarify when we expect `warp.UnpackSendWarpEventDataToMessage` to be successful and when we'd expect this fallback to be necessary?

It looks like `UnpackSendWarpEventDataToMessage` uses the `ParseUnsignedMessage` under the hood, but it passes `event.Message` to it, so I'm not sure which is the expected case.

May want to consider adding a debug log for the fallback case here also.

Reply:
In the expected case, `unsignedMsgBytes` is actually to abi encoded `SendMessage` event. This is parsed, and the `Message` field passed to `ParseUnsignedMessage`. 

The fallback case pertains only to manual Warp messages, in which `unsignedMsgBytes` is the Warp message itself, so can be directly passed to `ParseUnsignedMessage`.